procrastinate/worker.py

```diff
@@ -338,6 +337,49 @@ async def run(self):
                 pass
             raise
 
+    async def _handle_notification(
+        self, *, channel: str, notification: jobs.Notification
+    ):
+        if (
+            notification["type"] == "job_inserted"
+            and notification["status"] == jobs.Status.TODO.value
+        ):
+            self._new_job_event.set()
+        elif notification["type"] == "abort_job_requested":
+            if any(
+                job_context
+                for job_context in self._running_jobs.values()
+                if job_context.job.id == notification["job_id"]
+            ):
+                # cancellation notification is for one of the runnings jobs from this worker
+                self._job_ids_to_abort.add(notification["job_id"])
+
+    async def _poll_jobs_to_abort(self):
+        while True:
+            logger.debug(
+                f"waiting for {self.polling_interval}s before querying jobs to abort"
+            )
+            await asyncio.sleep(self.polling_interval)
+            if not self._running_jobs:
+                logger.debug("Not querying jobs to abort because no job is running")
+                continue
+            try:
+                job_ids = await self.app.job_manager.list_jobs_to_abort_async()
+                running_job_ids = {
+                    c.job.id for c in self._running_jobs.values() if c.job.id
+                }
+
+                self._job_ids_to_abort.update(running_job_ids.intersection(job_ids))
```

Comment:
I'm very slightly worried that there are 2 independent places where we react to changes in the abort job list. I think I'd feel slightly better if there was a single function (well, coroutine) handling this.

That said, while the work we do on `self._job_ids_to_abort` is seemingly identical (replace a set vs update a set), it will be different when we actually abort the tasks, because we'll need to identify what jobs should be aborted which we haven't aborted already.

So all in all... Maybe we'll see how to refactor when we add the abortion itself. But I'd prefer if we don't have 2 very related pieces of code that need to do roughly the same thing sitting hundreds of lines appart.

Another solution could be having a queue and a coroutine awaiting on that queue, but that marginally doesn't solve the issue that we're going to do slightly different things on the notification vs the poll.

Trying to put a brain cell on that. It's not blocking for merging this PR.

Reply:
I refactored it to consolidate the logic.